### PR TITLE
UserInfo embed should include how a user's activity compares to others

### PR DESCRIPTION
**Pull request by jrusbatch**
Adds information about how the user's activity compares to other users on the server, based on the average number of messages the user has sent per day over the past 30 days. A medal or trophy will be displayed next to their rank for users who are in or above the 95th percentile. 

capture

### CODE REVIEW

File: Modix.Data/Models/Core/GuildUserParticipationStatistics.cs [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+        public ulong UserId { get; set; }
+
+        public int Rank { get; set; }
+
+        public decimal AveragePerDay { get; set; }
+
+        public int Percentile { get; set; }
+
+        public string GetParticipationEmoji()
```
Comment (JakenVeina) [inline Modix.Data/Models/Core/GuildUserParticipationStatistics.cs]: This strikes me as a view-related concern, embedded in the data layer.

File: Modix.Data/Repositories/MessageRepository.cs [hunk 2]
```diff
@@ -86,6 +89,49 @@ public async Task<int> GetGuildUserMessageCount(ulong guildId, ulong userId, Tim
             return result.ToDictionary(x => userQuery[x.Key], x => x.Count());
         }
 
+        public async Task<GuildUserParticipationStatistics> GetGuildUserParticipationStatistics(ulong guildId, ulong userId)
+        {
+            var stats = await ModixContext.Database.GetDbConnection()
```
Comment (JakenVeina) [inline Modix.Data/Repositories/MessageRepository.cs]: Noooooow I see why you wanted me to look at this.

So. I'm not totally AGAINST using Dapper. But I think there needs to be a really strong argument for pulling in a SECOND Data Access library. And we need buy-in from all the primary maintainers.

I'll have a closer look at the query itself and see if I agree with you that (presumably) it's not doable in EF.

Comment (jrusbatch) [reply Modix.Data/Repositories/MessageRepository.cs]: I couldn't find a great way to express this with LINQ, and I didn't have a ton of luck figuring out how to run raw SQL (that wasn't associated with a `DbSet`) in EF Core, so I reverted to Dapper. Is there an easy way to do the same without Dapper?

File: Modix.Data/Models/Core/GuildUserParticipationStatistics.cs (hunk 1 above)
Comment (jrusbatch) [reply Modix.Data/Models/Core/GuildUserParticipationStatistics.cs]: Pushed 7111460f2d54e2fd613ae447dcb031da9c3a6e70, which moves this into `UserInfoModule`.

File: Modix.Data/Repositories/MessageRepository.cs (hunk 2 above)
Comment (Scott-Caldwell) [reply Modix.Data/Repositories/MessageRepository.cs]: After playing with it for a while, I was able to find some success with EF Core with the following changes. Not sure how idiomatic this is, but it :WorksOnMyMachine:
`MessageRepository.cs`: https://paste.mod.gg/ecobinavox.cs
`ModixContext.cs`: https://paste.mod.gg/iqasawemew.cs

File: Modix.Bot/Modules/UserInfoModule.cs [hunk 3]
```diff
@@ -198,11 +201,22 @@ private async Task AddParticipationToEmbedAsync(ulong userId, StringBuilder buil
...
+                builder.AppendFormat("Rank: {0} {1}\n", userRank.Rank.Ordinalize(), GetParticipationEmoji(userRank));
+            }
+
             builder.AppendLine("Last 7 days: " + weekTotal + " messages");
             builder.AppendLine("Last 30 days: " + monthTotal + " messages");
 
             if (monthTotal > 0)
             {
+                builder.AppendFormat(
+                    "Avg. per day: {0} messages (p{1})\n",
```
[the remark this answers is not in the document]
Comment (jrusbatch) [reply Modix.Bot/Modules/UserInfoModule.cs]: I was, too. I'm just not sure of a better way to express that. "(top __th percentile)" seemed a little verbose, but I could go back to that to try it out.

[the remark this answers is not in the document]
Comment (jrusbatch) [reply Modix.Bot/Modules/UserInfoModule.cs]: Updated.

File: Modix.Data/Repositories/MessageRepository.cs (hunk 2 above)
Comment (jrusbatch) [reply Modix.Data/Repositories/MessageRepository.cs]: Updated

**Pull request merged**